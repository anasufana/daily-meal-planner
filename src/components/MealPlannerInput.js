/*eslint class-methods-use-this: ["error", { "exceptMethods": ["validateCalories"] }] */

import React from 'react';
import '../css/components/MealPlannerInput.css';

class MealPlannerInput extends React.Component {
  constructor() {
    super();
    this.state = {
      targetCaloriesValue:
        {
          value: '',
          warning: true,
          warningText: 'Healthy target should be between 1000 and 3500',
        },
      excludeValue:
        {
          value: '',
          warning: false,
          warningText: 'Your exclude options are not valid. Try again.',
        },
      diets: [
        { value: '', name: 'Good to go!', selected: true },
        { value: 'gluten free', name: 'Gluten free', selected: false },
        { value: 'ketogenic', name: 'Ketogenic', selected: false },
        { value: 'vegetarian', name: 'Vegetarian', selected: false },
        { value: 'vegan', name: 'Vegan', selected: false },
        { value: 'pescetarian', name: 'Pescetarian', selected: false },
        { value: 'paleo', name: 'Paleo', selected: false },
      ],
    };

    this.handleTargetCalorieChange = this.handleTargetCalorieChange.bind(this);
    this.handleExcludeValue = this.handleExcludeValue.bind(this);
    this.validateCalories = this.validateCalories.bind(this);
  }
  validateCalories(calories) {
    if (calories) {
      if (calories >= 1000 && calories <= 3500) {
        return false;
      }
    }
    return true;
  }

  handleTargetCalorieChange(e) {
    const targetCalories = e.target.value;
    const warning = this.validateCalories(targetCalories);
    this.setState({
      targetCaloriesValue: { ...this.state.targetCaloriesValue, value: e.target.value, warning }
    });
  }

  handleExcludeValue(e) {
    const excludeText = e.target.value;
    let warning = false;

    //  Tests if the exclude text includes only characters, commas and spaces
    if (excludeText && !/^[a-zA-Z ,]+$/.test(excludeText)) {
      warning = true;
    }
    this.setState({ excludeValue: { ...this.state.excludeValue, value: e.target.value, warning } });
  }


  render() {
    return (
      <div className="meal-planner-input">
        <form onSubmit={(e) => { e.preventDefault(); }}>
          <div className="input-container">
            <span className="form-span">Choose your diet</span>
            <select className="dropdown-select">
              {this.state.diets.map(diet => (
                <option key={diet.value} value={diet.value} defaultValue={diet.selected}>
                  { diet.name }
                </option>
              ))
              }
            </select>
          </div>
          <div className="input-container">
            <span className="form-span">Exclude</span>
            <input
              className={this.state.excludeValue.warning ?
             'form-input-warning form-exclude' : 'form-input-text form-exclude'}
              type="text"
              value={this.state.excludeValue.value}
              onChange={this.handleExcludeValue}
            /><br />
            <span className={this.state.excludeValue.warning ? 'form-input-info-warning' : 'display-none'}>
              {this.state.excludeValue.warningText}
            </span>
            <span className="form-input-info">
              Example: nuts, shallots - separated by comma.<br />
              Leave blank if you don&apos;t want to exclude anything
            </span>
          </div>
          <div className="input-container input-calories-container">
            <span className="form-span">Target calories*</span>
            <input
              className={this.state.targetCaloriesValue.warning
               ? 'form-input-warning form-calories' : 'form-input-text form-calories'}
              type="text"
              value={this.state.targetCaloriesValue.value}
              onChange={this.handleTargetCalorieChange}
            /><br />
            <span className={this.state.targetCaloriesValue.warning
               ? 'form-input-info-warning' : 'form-input-info'}
            >
              {this.state.targetCaloriesValue.warningText}
            </span>
          </div>
          <button className="submit-btn"type="submit" value="Submit" disabled={(this.state.targetCaloriesValue.warning || this.state.excludeValue.warning)}>Go!</button>
        </form>
      </div>
    );
  }
}

export default MealPlannerInput;
