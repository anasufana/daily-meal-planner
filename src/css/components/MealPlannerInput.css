.meal-planner-input {
  width: 600px;
  margin: 2rem auto;
  padding: 2rem 0;
  border-top: 2px solid black; }

.input-container {
  width: 100%;
  position: relative;
  text-align: right;
  padding: 0.75rem 0; }
  .input-container .form-span {
    margin-right: 1.5rem;
    font-size: 1.6rem; }
  .input-container .dropdown-select, .input-container .form-input-text, .input-container .form-input-warning {
    width: 300px;
    height: 50px;
    border-radius: 5px;
    border: 2px solid black;
    transition: 0.25s all ease-in;
    font-size: 1rem;
    padding: 0.5rem;
    font-family: 'Nanum Gothic', sans-serif; }
    .input-container .dropdown-select:focus, .input-container .form-input-text:focus, .input-container .form-input-warning:focus {
      outline: none;
      border: 2px solid pink; }
  .input-container .form-input-warning:focus {
    border: 2px solid #a70505; }
  .input-container .form-input-info, .input-container .form-input-info-warning {
    font-size: 0.75rem;
    margin-top: 0.25rem;
    padding-left: 310px;
    display: block;
    text-align: left;
    trnasition: 0.3s all ease-in; }
  .input-container .form-input-info-warning {
    color: #a70505; }

.submit-btn {
  height: 50px;
  width: 300px;
  border-radius: 5px;
  color: white;
  border: 2px solid black;
  background-color: black;
  padding: 0.5rem 3rem;
  float: right;
  margin-top: 0.75rem;
  transition: 0.25s all ease-in;
  font-family: 'Nanum Gothic', sans-serif;
  font-size: 1rem;
  cursor: pointer; }
  .submit-btn:focus, .submit-btn:hover {
    outline: none;
    background-color: white;
    color: black; }
  .submit-btn:focus {
    border: 2px solid pink; }
  .submit-btn:disabled, .submit-btn[disabled] {
    background-color: #7d7d7d;
    color: black;
    border: 2px solid #7d7d7d;
    cursor: default; }
