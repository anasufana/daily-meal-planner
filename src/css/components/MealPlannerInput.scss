.meal-planner-input {
  width: 550px;
  margin: 4rem auto;
  padding: 2rem 0;
  border-top: 2px solid black;
}

.input-container {
  width:100%;
  position: relative;
  text-align: left;
  padding: 0.75rem 0;
}

.form-span {
    margin-right: 1.5rem;
    font-size: 1.6rem;
    margin-bottom: 10px;
}

.dropdown-select, .form-input-text, .form-input-warning {
  width: 100%;
  height: 50px;
  border-radius: 5px;
  border: 2px solid black;
  transition: 0.25s all ease-in;
  font-size: 1rem;
  padding: 0.5rem;
  margin-top: 10px;
  font-family: 'Nanum Gothic', sans-serif;

  &:focus {
    outline: none;
    border: 2px solid pink;
  }
}

  .form-input-warning {
    &:focus {
      border: 2px solid #a70505;
    }
  }

  .form-input-info, .form-input-info-warning {
    font-size: 0.75rem;
    margin-top: 0.25rem;
    display: block;
    text-align: left;
    trnasition: 0.3s all ease-in;
  }

  .form-input-info-warning {
    color: #a70505;
  }


.submit-btn {
  height: 50px;
  width: 100%;
  border-radius: 5px;
  color: white;
  border: 2px solid black;
  background-color: black;
  padding: 0.5rem 3rem;
  float: right;
  margin-top: 0.75rem;
  transition: 0.25s all ease-in;
  font-family: 'Nanum Gothic', sans-serif;
  font-size: 1rem;
  cursor: pointer;

  &:focus, &:hover {
    outline: none;
    background-color: white;
    color: black;
  }

  &:focus {
    border: 2px solid pink;
  }

  &:disabled, &[disabled] {
    background-color: rgb(125,125,125);
    color: black;
    border: 2px solid rgb(125,125,125);
    cursor: default;
  }

}


.display-none {
  display: none;
}

@media screen and (max-width: 769px) {
  .meal-planner-input {
    margin-top: 1em;
  }
}
