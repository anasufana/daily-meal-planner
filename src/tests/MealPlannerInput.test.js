/* global it, expect, jest */

import React from 'react';
import renderer from 'react-test-renderer';
import MealPlannerInput from '../components/MealPlannerInput';

const state = {
  targetCaloriesValue:
    {
      value: 'banana',
      warning: false,
      warningText: 'Healthy target should be between 1000 and 3500',
    },
  excludeValue:
    {
      value: '2000',
      warning: false,
      warningText: 'Your exclude options are not valid. Try again.',
    },
  diets: [
    { value: '', name: 'Good to go!', selected: false },
    { value: 'gluten free', name: 'Gluten free', selected: false },
    { value: 'ketogenic', name: 'Ketogenic', selected: false },
    { value: 'vegetarian', name: 'Vegetarian', selected: true },
    { value: 'vegan', name: 'Vegan', selected: false },
    { value: 'pescetarian', name: 'Pescetarian', selected: false },
    { value: 'paleo', name: 'Paleo', selected: false },
  ],
  disableButton: false,
};

it('renders correctly', () => {
  const tree = renderer
    .create(<MealPlannerInput
      {...state}
      handleSubmit={jest.fn()}
      handleDietSelect={jest.fn()}
      handleExcludeValue={jest.fn()}
      handleTargetCalorieChange={jest.fn()}
    />).toJSON();
  expect(tree).toMatchSnapshot();
});
