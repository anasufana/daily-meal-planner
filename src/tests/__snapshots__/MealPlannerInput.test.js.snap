// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="meal-planner-input"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="input-container"
    >
      <span
        className="form-span"
      >
        Choose your diet
      </span>
      <select
        className="dropdown-select"
        onChange={[Function]}
      >
        <option
          defaultValue={false}
          value=""
        >
          Good to go!
        </option>
        <option
          defaultValue={false}
          value="gluten free"
        >
          Gluten free
        </option>
        <option
          defaultValue={false}
          value="ketogenic"
        >
          Ketogenic
        </option>
        <option
          defaultValue={true}
          value="vegetarian"
        >
          Vegetarian
        </option>
        <option
          defaultValue={false}
          value="vegan"
        >
          Vegan
        </option>
        <option
          defaultValue={false}
          value="pescetarian"
        >
          Pescetarian
        </option>
        <option
          defaultValue={false}
          value="paleo"
        >
          Paleo
        </option>
      </select>
    </div>
    <div
      className="input-container"
    >
      <span
        className="form-span"
      >
        Exclude
      </span>
      <input
        className="form-input-text form-exclude"
        onChange={[Function]}
        type="text"
        value="2000"
      />
      <br />
      <span
        className="display-none"
      >
        Your exclude options are not valid. Try again.
      </span>
      <span
        className="form-input-info"
      >
        Example: nuts, shallots - separated by comma.
        <br />
        Leave blank if you don't want to exclude anything
      </span>
    </div>
    <div
      className="input-container input-calories-container"
    >
      <span
        className="form-span"
      >
        Target calories*
      </span>
      <input
        className="form-input-text form-calories"
        onChange={[Function]}
        type="text"
        value="banana"
      />
      <br />
      <span
        className="form-input-info"
      >
        Healthy target should be between 1000 and 3500
      </span>
    </div>
    <button
      className="submit-btn"
      disabled={false}
      type="submit"
      value="Submit"
    >
      Go!
    </button>
  </form>
</div>
`;
